javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
> In other words your version is not deterministic and may produce unexpected results.

Worse than that: reading a naked `value` is reading via the race. There is no guarantees that something that you have written **into the value fields themselves** is visible to the readers of value reference. Should read supplier first to get "acquire" part.